bin/BuildPackages.sh

```diff
@@ -81,60 +92,75 @@ build_carat() {
 zcat carat-2.1b1.tgz | tar pxf -
 ln -s carat-2.1b1/bin bin
 cd carat-2.1b1
-make TOPDIR=`pwd`
+make TOPDIR="$(pwd)"
 chmod -R a+rX .
 cd bin
-aa=`./config.guess`
-for x in "`ls -d1 $GAPDIR/bin/${aa}*`"; do
- ln -s "$aa" "`basename $x`"
+aa=$(./config.guess)
+# TODO dangerous to parse output of ls, use find instead
+for x in "$(ls -d1 $GAPDIR/bin/${aa}*)"
+do
+  ln -s "$aa" "$(basename $x)"
 done
 )
 }
 
 build_cohomolo() {
 (
-./configure $GAPDIR
+./configure "$GAPDIR"
 cd standalone/progs.d
 cp makefile.orig makefile
 cd ../..
-$MAKE
+"$MAKE"
 )
 }
 
 build_fail() {
-  echo "= Failed to build $dir"
+  echo ""
+  echo "=* Failed to build $dir"
+  echo "$dir" >> "$LOGDIR/$FAILPKGFILE.log"
 }
 
 run_configure_and_make() {
   # We want to know if this is an autoconf configure script
   # or not, without actually executing it!
-  if [ -f autogen.sh ] && ! [ -f configure ] ; then
+  if [[ -f autogen.sh ]] && ! [[ -f configure ]]
+  then
     ./autogen.sh
-  fi;
-  if [ -f configure ]; then
-    if grep Autoconf ./configure > /dev/null; then
-      ./configure $CONFIGFLAGS --with-gaproot=$GAPDIR
+  fi
+  if [[ -f "configure" ]]
+  then
+    if grep Autoconf ./configure > /dev/null
+    then
+      ./configure "$CONFIGFLAGS" --with-gaproot="$GAPDIR"
     else
-      ./configure $GAPDIR
-    fi;
-    $MAKE
+      ./configure "$GAPDIR"
+    fi
+    "$MAKE"
   else
     echo "No building required for $dir"
-  fi;
+  fi
 }
 
-for dir in `ls -d */`
+date >> "$LOGDIR/$FAILPKGFILE.log"
+for dir in $(find -L . -maxdepth 1 -mindepth 1 -type d)
 do
-    if [ -e $dir/PackageInfo.g ]; then
-      dir="${dir%/}"
+  if [[ -e "$dir/PackageInfo.g" ]]
+  then
+    dir="${dir%/}"
+    dir="${dir##*/}"
+
+    buildpackage() {
```

Comment:
another advantage of this is that we are then very close to being able to build just a single package -- say, specified via an argument to the script.

I.e. the script could have some actual argument parsing at the start, accepting an `--with-gaproot=DIR` argument, optionally followed by a list of package names (or names of package dirs) to build...